### DISCUSSION TITLE
Add custom co-visualizations to the tree

### DISCUSSION THREAD

**Discussion by pentamorfico** [theosanderson/taxonium#476] [Ideas]
Hi Theo! first of all congratulations to you and the other contributors for this great tool!

I'm fascinated with the fluidity and ease of handling such huge trees! This is my new fav tool to analyze large phylogenies for sure!!

I was wondering if it would be somehow possible to co-visualize the tree with other features beyond JBrowse. I was thinking for example heatmaps, or gene-arrow diagrams (see https://github.com/wilkox/gggenes for an example) for comparative genomics purposes.

I have been playing around a bit with deck.gl, and kind of managed to make a gene arrow-like diagram, but do you think it would it be easy to implement this kind of co-visualizations with the tree?? It would take the field of comparative genomics visualization to another level!

**Comment by theosanderson** [comment]
Hello! Thank you for making the first Discussion on this repo and sorry that it's taken so kong to reply. Thanks for the kind words. This is an interesting concept. One of the reasons I've taken a bit of time to reply is that I didn't fully get my head around the concept. Were you thinking about gene arrows to show synteny or something? I think my questions would be partly about how things would be displayed when zoomed far out. I will definitely keep these ideas in mind as development continues. Feel free to be a bit more explicit about what you would imagine if you want (though I still can't promise to deliver).

Sorry again for the slow reply and thank you for the feedback.

**Reply by pentamorfico** [reply]
Hi Theo, thank you for taking the time to read this! I really appreciate it!

So, my thoughts were about using taxonium to visualize large phylogenies, not necessarily limited to SARS-CoV-2. The idea is to visualize gene synteny along with the phylogenetic tree. In the context of bacterial comparative genomics, it would be fascinating to analyze various elements such as biosynthetic gene clusters, antibiotic resistance genes, and mobile genetic elements across extensive datasets. The main concept would involve displaying gene arrows similar to gggenes or GCsnap (https://github.com/JoanaMPereira/GCsnap) when zoomed in. As you say, when zoomed out, it would be challenging to visualize these gene arrows. In such cases, switching to a heatmap or a presence/absence matrix demonstrating the presence or absence of operons could be highly beneficial. You can refer to this link (https://github.com/visgl/deck.gl/discussions/7481) for an example.

I understand that implementing this for whole genomes could be computationally expensive, but it might be a great starting point and extremely useful for visualizing operons or regions around specific genes.

Does this idea make sense, or is it beyond the scope of this tool?

Cheers,
Mario

**Comment by lm-jkominek** [comment]
Hi, @theosanderson, first of all - Taxonium is very impressive and works great, love it! I wanted to upvote and contribute to this discussion because I had the exact same thought when I came across Taxonium (almost created an issue for it, but thankfully thought to check discussions first). 
 
My use case is similar to @pentamorfico - I have a large static tree of a set of genomes and would like to annotate it with various comparative genomics features - mainly presence/absence of genes, but also potentially heatmaps and synteny plots. The first two, in principle, can be done with iTOL datasets (e.g. https://itol.embl.de/help.cgi#binary) but iTOL's performance is terrible at anything above 1,000 tips. In terms of the synteny plots I was envisioning something like Clinker plots (https://github.com/gamcil/clinker).

To respond to your question about zoom levels, I would imagine using some form of data condensation/simplification at deeper nodes and showing that when zoomed out. This processing could be either hardcoded (e.g. show the most common data in a clade, or the average of it, the largest/longest data point or something else) or it can be provided by the user, tying custom data to internal nodes of the tree.

I hope that helps, happy to discuss this more, if there's a chance that this feature would make it to Taxonium!

**Reply by theosanderson** [reply]
Thanks to you and @pentamorfico (to whom I really owe a reply, very sorry!) for raising this. It's really useful to know that a minimum of two people are looking for this, and others are welcome to chime in. I think this may be unlikely to come to Taxonium soon just because of the limitations of manpower (as you can see from my failures to reply!) and possibly also to avoid the package becoming unwieldy. One thing that we should try to do is to improve `taxonium-component`, our reusable react component, such that people are empowered to build something that puts Taxonium next to such a feature display.

**Reply by lm-jkominek** [reply]
Thanks @theosanderson, totally understandable! I'm with you, and if there would be a way to utilize `taxonium-component` for this purpose, even through extra custom code on the user's side, that would be fantastic.

**Comment by pentamorfico** [comment]
Hi @theosanderson no worries about the delay! And no worries, I completely understand your situation! I would love to have the necessary experience with javascript to be able to help but at the moment I only use it for very basic things! if in the future I do something that can fit I will let you know!

In the meantime, the developers of ete-toolkit for Python have release this amazing tool that may fit the needs that we were talking about: https://github.com/compgenomicslab/TreeProfiler Check it out @lm-jkominek, I think you are going to like it!

Cheers, 
Mario

**Reply by lm-jkominek** [reply]
Thanks a lot, @pentamorfico! Haven't tested it yet, but the TreeProfiler does look great and very promising, so if the interactive navigation works as well as I hope, it might just fill that need we talked about here...

**Comment by pentamorfico** [comment]
Hi @lm-jkominek @theosanderson! After quite a bit of tinkering (and many lessons learned along the way), I put together a small project that might help explore this space:

https://pentamorfico.github.io/hoodini-docs/

It’s definitely not as scalable as Taxonium (and is more centered on gene exploration than tree exploration), but I hope it can serve as a starting point or a source of ideas. Building it gave me some practical insight into how a similar approach could potentially be implemented within Taxonium, and I’d be very happy to discuss, iterate on it, or adapt ideas together.
